Initial implementation of rolling iterators

ThomasKluiters opened the pull request:
- [x] closes #11704
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Hi, I have yet to implements tests, use flake8 and create a whatsnew entry as I am uncertain whether this implementation would be correct.

I would really appreciate it if someone could give me some feedback on the code.

This PR will implement iterators for window functions.

Some questions I have:

* Does DataFrame.rolling(x) yield DataFrames?
* Does Series.rolling(x) yield Series?

ThomasKluiters committed:
Initial implementation of rolling iterators

pep8speaks commented:
Hello @ThomasKluiters! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/window.py`](https://github.com/pandas-dev/pandas/blob/0cd558eb8ddf6a91789554ed0fcca63137866d14/pandas/core/window.py):

> [Line 220:89](https://github.com/pandas-dev/pandas/blob/0cd558eb8ddf6a91789554ed0fcca63137866d14/pandas/core/window.py#L220): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)
> [Line 221:1](https://github.com/pandas-dev/pandas/blob/0cd558eb8ddf6a91789554ed0fcca63137866d14/pandas/core/window.py#L221): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

WillAyd commented:
Can you add tests for this?

WillAyd commented:
Also I think yes to both of your questions

ThomasKluiters commented:
> Can you add tests for this?

I wouldn't mind adding tests! I had to know what the expected output would be, but now that I know I'll be writing them! Thanks :)

ThomasKluiters committed:
Implement simple test cases

ThomasKluiters committed:
Add whatsnew entry

ThomasKluiters commented:
@WillAyd Hey, I added some simple test cases, I don't know how much test cases will be required, as there's a significant amount of functionality to cover. I will be adding some more test cases for Date aware indices. Could you have a look what I have so far :) ?

ThomasKluiters committed:
Fix Cython compile error

ThomasKluiters committed:
Add expanding test cases

ThomasKluiters committed:
Implement date time index aware window

ThomasKluiters committed:
Reformat

ThomasKluiters committed:
Format window.pyx

mroeschke commented on pandas/core/window.py:
```diff
@@ -204,8 +204,21 @@ def __repr__(self):
+        blocks, obj, index = self._create_blocks()
+        index, indexi = self._get_index(index)
+
+        iterators = []
```
Do we need to accumulate the iterators in a list? Can we just `yield`?

ThomasKluiters replied:
We could iterate over each iterator and yield each element, though, this might diminish any performance we've gained by implementing our own Iterator.

Would you envision something like this?

```python
for block in blocks:
   for item in iterator_for_block(block):
      yield item
```

We could also lazily create the new iterators

mroeschke replied:
I was thinking:

```
for values in blocks:
    for element in libwindow.WindowIterator(values, indexi, window, self.closed, minp):
        yield element
```

ThomasKluiters replied:
Yeah exactly :)

I shortened it a bit as I didn't have the source code available.

mroeschke commented on pandas/core/window.py:
```diff
@@ -204,8 +204,21 @@ def __repr__(self):
     def __iter__(self):
-        url = "https://github.com/pandas-dev/pandas/issues/11704"
-        raise NotImplementedError("See issue #11704 {url}".format(url=url))
+        self.validate()
```
`self.validate()` is already done in the `__init__`

ThomasKluiters replied:
Well spotted! Will remove :)

mroeschke commented on pandas/_libs/window.pyx:
```diff
@@ -354,6 +354,42 @@ def get_window_indexer(values, win, minp, index, closed,
     return indexer.get_data()
 
 
+cdef class WindowIterator:
```
Not opposed to using cython, but curious why this can't be implemented directly on the `class _Window` base class?

ThomasKluiters replied:
It can be implemented directly, though, we'd have to call libwindow's `get_indexer`, I thought creating this iterator inside the cython file might be better for that reason.

mroeschke replied:
I think it should be okay to import `get_indexer` directly and use it to slice the `DataFrames`/`Series`

ThomasKluiters replied:
Done!

I did some performence tests and the performance loss is not significant. (A loss of around ~5%)

mroeschke commented on pandas/tests/window/test_expanding.py:
```diff
@@ -90,13 +90,60 @@ def test_missing_minp_zero(self):
+        ],
+    )
+    def test_iterator_dataframe(self, dataframe, expected, window):
+        expected = [DataFrame(values, index=index) for (values, index) in expected]
```
The parametrize signature should probably be `"dataframe,expected_data,index,window"`

Then the test body can be:

```
expected = DataFrame(expected_data, index=index)
result = dataframe.expanding(window)
tm.assert_frame_equal(actual, expected)
```

ThomasKluiters replied:
The Index and values are different for each iteration, so, for each item part of the iterator we'd have a pair of (expected_values, expected_index).

mroeschke replied:
Got it, thanks for the explanation. Fine as is then.

Commit:
Fix failing tests

ThomasKluiters committed:
Remove validate from iter

ThomasKluiters committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into rolling-iterator

ThomasKluiters committed:
Refactor get_window_indexer to _Window

ThomasKluiters committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into rolling-iterator

ThomasKluiters committed:
Merge branch 'rolling-iterator' of https://www.github.com/ThomasKluiters/pandas into rolling-iterator

ThomasKluiters commented:
@mroeschke I have changed the files according to your feedback, could you have a look?

mroeschke commented on pandas/core/window.py:
```diff
@@ -204,8 +204,27 @@ def __repr__(self):
+        blocks, obj, index = self._create_blocks()
+        _, indexi = self._get_index(index)
+
+        for _, values in enumerate(blocks):
```
Don't need the `enumerate` here

mroeschke reviewed (changes requested):
Looking good. Could you also add tests where `min_periods` is utilized?

WillAyd reviewed (changes requested):
Can you add `expanding` to the tests here as well?

ThomasKluiters committed:
Remove enumerate

ThomasKluiters commented:
> Can you add `expanding` to the tests here as well?

They're already there, view the changes under test_expanding :)

WillAyd commented:
@ThomasKluiters thanks - sorry misread that on my end

ThomasKluiters committed:
Add tests for min_periods

ThomasKluiters commented:
> Looking good. Could you also add tests where `min_periods` is utilized?

@mroeschke 
Expanding kind of already does this, but I have added three new unit tests that use window and min_periods, please check!

ThomasKluiters committed:
Reformat tests

mroeschke commented on pandas/tests/window/test_rolling.py:
```diff
@@ -98,6 +98,95 @@ def test_constructor_timedelta_window_and_minperiods(self, window, raw):
+        "dataframe,expected,window",
+        [
+            (
+                DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]}),
```
The `dataframe` and `series` parameters all look to be the same across the tests. You can bring them into the test directly (will be easier to see what's being

mroeschke replied:
Maybe not this test specifically, but the ones where applicable.

mroeschke commented:
For the `min_periods` tests, could you also test where there's `nan`s in the Series/DataFrames?

ThomasKluiters committed:
Refactor tests

ThomasKluiters committed:
Implement support for nan values

ThomasKluiters commented:
> For the `min_periods` tests, could you also test where there's `nan`s in the Series/DataFrames?

Well, it turns out that this required a bit of work - I've added some tests for it and had to change some code in window.py

ThomasKluiters committed:
Rename duplicate test name

mroeschke commented on pandas/core/window.py:
```diff
@@ -204,8 +204,47 @@ def __repr__(self):
+            if arr.ndim == 1:
+                arr = np.expand_dims(arr, axis=1)
+
+            counts = libwindow.roll_sum(
```
Instead of this, could you just generate the slice and yield if the number non-null elements meets `_minp`?

```
result = values.iloc[slice(s, e)]
if result.count() > _minp:
    yield result
```

ThomasKluiters committed:
Use count() instead of roll_sum()

ThomasKluiters committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into rolling-iterator

ThomasKluiters committed:
Update create_blocks and _get_index

mroeschke reviewed (approved):
Looks pretty good! Over to @WillAyd, @jreback for a second look.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -1206,6 +1206,7 @@ Groupby/resample/rolling
 - Improved :class:`pandas.core.window.Rolling`, :class:`pandas.core.window.Window` and :class:`pandas.core.window.EWM` functions to exclude nuisance columns from results instead of raising errors and raise a ``DataError`` only if all columns are nuisance (:issue:`12537`)
 - Bug in :meth:`pandas.core.window.Rolling.max` and :meth:`pandas.core.window.Rolling.min` where incorrect results are returned with an empty variable window (:issue:`26005`)
 - Raise a helpful exception when an unsupported weighted window function is used as an argument of :meth:`pandas.core.window.Window.aggregate` (:issue:`26597`)
+- Improved :class:`pandas.core.window.Rolling` and :class:`pandas.core.window.Window` it is now possible to iterate over a window object (:issue:`12537`)
```
move to 1.0 whatsnew

jreback replied:
also provide a reference to the new doc section.

jreback commented on pandas/tests/window/test_expanding.py:
```diff
@@ -90,13 +90,62 @@ def test_missing_minp_zero(self):
+        ],
+    )
+    def test_iterator_dataframe(self, dataframe, expected, window):
+        expected = [DataFrame(values, index=index) for (values, index) in expected]
```
can you add the issue number as a comment here

jreback commented on pandas/tests/window/test_expanding.py:
```diff
@@ -90,13 +90,62 @@ def test_missing_minp_zero(self):
+        ],
+    )
+    def test_iterator_series(self, series, expected, window):
+        expected = [Series(values, index=index) for (values, index) in expected]
```
same

jreback reviewed (changes requested):
can you add a small section in computation.rst which shows off the iterator, probably in the rolling parts would work well. 

I had a thought to co-locate all of these tests in pandas/tests/window/test_iterator.py (rather than splitting them off to the sub-files). I am not convinced our current organziation of tests here is that great. 

otherwise lgtm.

ThomasKluiters committed:
Refactor tests into test_iterator

ThomasKluiters committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into rolling-iterator

ThomasKluiters committed:
Move whatsnew entry to 1.0.0

ThomasKluiters committed:
Add iterators to computation.rst

ThomasKluiters committed:
Add issue #

ThomasKluiters commented:
@jreback 

I have moved the tests to `test_iterator` and added the reference to the issue at the top.
I have moved the whatsnew to 1.0.0.
I have added the documentation to computation.rst

Though, I can't seem to figure out how to reference computation.rst in the Whatsnew, could you help me with this please :)? Thanks!

jreback commented:
can you merge master and see if can get passing

jreback commented:
IIRC this looked reasonable, can you merge master and update to any comments

jreback commented:
closing this as stale, though its a nice change, needs a rebase and comments addressed.
